src/object/tcb.c

```diff
@@ -1251,22 +1228,89 @@ exception_t decodeSetMCPriority(cap_t cap, word_t length, word_t *buffer)
 }
 
 #ifdef CONFIG_KERNEL_MCS
-exception_t decodeSetTimeoutEndpoint(cap_t cap, cte_t *slot)
+static bool_t validFaultHandler(cap_t cap)
+{
+    switch (cap_get_capType(cap)) {
+    case cap_endpoint_cap:
+        if (!cap_endpoint_cap_get_capCanSend(cap) ||
+            (!cap_endpoint_cap_get_capCanGrant(cap) &&
+             !cap_endpoint_cap_get_capCanGrantReply(cap))) {
+            current_syscall_error.type = seL4_InvalidCapability;
+            return false;
+        }
+        break;
+    case cap_null_cap:
+        /* just has no fault endpoint */
+        break;
+    default:
+        current_syscall_error.type = seL4_InvalidCapability;
+        return false;
+    }
+    return true;
+}
+
+static deriveCap_ret_t updateAndCheckHandlerEP(cap_t cap, word_t length, cte_t *slot, word_t *buffer,
+                                               word_t offset, word_t capPosition)
+{
```

Comment:
Add sanity check for catching development errors and clarifying function calling contract.
```suggestion
{
    assert( !(buffer && (length > 0) );
    assert( (0 == length) || (offset < length) );
```